File: app_old/app/src/main/java/com/godot17/arksc/activity/ScreenshotActivity.java
package com.godot17.arksc.activity;

import static com.godot17.arksc.activity.ScreenTaskActivity.getTagText;
import static com.godot17.arksc.utils.ScreenUtils.getDensityDpi;
import static com.godot17.arksc.utils.ScreenUtils.getRealHeight;
import static com.godot17.arksc.utils.ScreenUtils.getRealWidth;
import static com.godot17.arksc.utils.Utils.getAssets2CacheDir;

import static java.lang.Thread.sleep;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.PixelFormat;
import android.hardware.display.DisplayManager;
import android.hardware.display.VirtualDisplay;
import android.media.Image;
import android.media.ImageReader;
import android.media.projection.MediaProjection;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.Surface;
import android.widget.Toast;
import android.widget.Toolbar;

import androidx.annotation.Nullable;

import com.godot17.arksc.App;
import com.godot17.arksc.database.NewOpe;
import com.godot17.arksc.service.DataProcessService;
import com.godot17.arksc.utils.ImageUtils;

public class ScreenshotActivity extends Activity {
    public static final String TAG = "ScreenTaskActivity";
    private MediaProjection mediaProjection;
    private static ImageReader imageReader;
    private Surface surface = null;
    private VirtualDisplay virtualDisplay;
    private int screenWidth;
    private int screenHeight;
    private int screenDensityDpi;

    private Handler mHandler;


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        new Thread(() -> {
            Looper.prepare();
            mHandler = new Handler();
            Looper.loop();
        }).start();
        super.onCreate(savedInstanceState);
    }

    @Override
    protected void onStart() {
        Log.e(TAG,"ONSTART");
        setScreenSize();
        prepareForScreenshot();
        takeOneScreenshot();
        super.onStart();
    }

    private void setScreenSize() {
        screenWidth = getRealWidth(this);
        screenHeight = getRealHeight(this);
        screenDensityDpi = getDensityDpi(this);
    }

    private void prepareForScreenshot() {
        try {
            if (imageReader == null)
                imageReader = ImageReader.newInstance(screenWidth, screenHeight,
                        PixelFormat.RGBA_8888, 2);
            surface = imageReader.getSurface();
            mediaProjection = App.createMediaProjection();
            startVirtualDisplay();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void startVirtualDisplay() {
        if (virtualDisplay != null) virtualDisplay.release();
        virtualDisplay = mediaProjection.createVirtualDisplay(
                "ScreenShot",
                screenWidth,
                screenHeight,
                screenDensityDpi,
                DisplayManager.VIRTUAL_DISPLAY_FLAG_OWN_CONTENT_ONLY,
                surface,
                null,
                null
        );
    }

    private void takeOneScreenshot() {
        Thread thread = new Thread(() -> {
            try {
                sleep(300);
                Image image = imageReader.acquireLatestImage();
                Bitmap bitmap;
                String[] data;

                bitmap = ImageUtils.image_2_bitmap(image, Bitmap.Config.ARGB_8888);
                bitmap = ImageUtils.bitmap_2_target_bitmap(bitmap, screenWidth, screenHeight);
                bitmap.recycle();
                int scale = ImageUtils.getScale(screenWidth);
                String dataPath = getAssets2CacheDir(this, "target_std.dat");
                data = getTagText(bitmap, dataPath, scale).split(",");
                switch (data[0]) {
                    case "NONE":
                        throw new Exception("未获取有效信息");
                    case "WRONG":
                        throw new Exception(data[1]);
                    case "RECRUIT": {
                        Log.e(TAG, "RECRUIT");
                        //TODO:
                        Intent dataProcess = new Intent(this, DataProcessService.class);
                        dataProcess.putExtra("TAG_SORT", "RECRUIT");
                        dataProcess.putExtra("TAG_TEXT", data[1]);
                        this.startService(dataProcess);
                        break;
                    }
                }
                finish();
            } catch (Exception e) {
                mHandler.post(() -> Toast.makeText(this, e.getMessage(), Toast.LENGTH_LONG).show());
            }
        });
        thread.start();
    }

    @Override
    protected void onDestroy() {
        stop();
        super.onDestroy();
    }

    private void stop() {
        stopVirtualDisplay();
        stopSurface();
        imageReader = null;
        stopMediaProjection();
        releaseMutex();
    }

    private void stopMediaProjection() {
        if (mediaProjection != null) mediaProjection.stop();
        mediaProjection = null;
    }

    private void stopVirtualDisplay() {
        if (virtualDisplay != null) virtualDisplay.release();
        virtualDisplay = null;
    }

    private void stopSurface() {
        if (surface != null) surface.release();
        surface = null;
    }

    private void releaseMutex() {
        App.releaseMutexActivity();
    }
}

File: app/src/main/java/com/blueskybone/arkscreen/task/screenshot/ScreenshotActivity.kt
package com.blueskybone.arkscreen.task.screenshot

import android.app.Activity
import android.graphics.Bitmap
import android.graphics.PixelFormat
import android.hardware.display.DisplayManager
import android.hardware.display.VirtualDisplay
import android.media.ImageReader
import android.media.projection.MediaProjection
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.Surface
import android.widget.Toast
import com.blueskybone.arkscreen.APP
import com.blueskybone.arkscreen.I18n
import com.blueskybone.arkscreen.base.PrefManager
import com.blueskybone.arkscreen.task.recruit.RecruitManager
import com.blueskybone.arkscreen.util.convertImageToBitmap
import com.blueskybone.arkscreen.util.getDensityDpi
import com.blueskybone.arkscreen.util.getEleCombination
import com.blueskybone.arkscreen.util.getRealScreenSize
import com.hjq.toast.Toaster
import org.koin.android.ext.koin.androidLogger
import org.koin.core.context.startKoin
import org.koin.dsl.module
import org.koin.java.KoinJavaComponent.getKoin

/**
 *   Created by blueskybone
 *   Date: 2024/7/26
 */

class ScreenshotActivity : Activity() {

    private val prefManager: PrefManager by getKoin().inject()

    companion object {
        private const val TAG = "ScreenshotActivity"

        private var mediaProjection: MediaProjection? = null
        private var imageReader: ImageReader? = null
        private var surface: Surface? = null
        private var virtualDisplay: VirtualDisplay? = null

        var screenWidth = 0
        var screenHeight = 0
        var screenDensityDpi = 0
        private const val delay: Long = 300

        var debug: Boolean = false
    }

    private fun setScreenSize() {
        screenDensityDpi = getDensityDpi(this)
        val point = getRealScreenSize(this)
        screenWidth = point.x
        screenHeight = point.y

//        if (debug)
//            loggerRun.write("screenWidth: $screenWidth  screenHeight: $screenHeight")
    }


    override fun onStart() {
        super.onStart()
        debug = prefManager.enableDebug.get()
        try {
            setScreenSize()
            if (prepareForScreenshot()) {
                val bitmap = takeOneScreenshot()
                if (bitmap != null) {
                    Thread {
                        processBitmap(bitmap)
                    }.start()
                }
            }
            stop()
        } catch (e: Exception) {
//            loggerTrace.write("$TAG: $e \n${exceptionStackTrace(e)}")
            stop()
        }
    }

    private fun takeOneScreenshot(): Bitmap? {
        Thread.sleep(delay)
        var bitmap: Bitmap? = null
//        if (debug)
//            loggerRun.write("takeOneScreenshot")
        try {
            imageReader!!.acquireLatestImage().use { image ->
                if (image != null) {
                    bitmap = convertImageToBitmap(image, Bitmap.Config.ARGB_8888)
//                    if (debug) {
//                        loggerRun.write("convertImageToBitmap")
//                        saveBitmap(bitmap!!, "raw_bitmap.jpg")
//                        loggerRun.write("save raw screenshot")
//                    }
                } else {
//                    loggerTrace.write("$TAG: takeOneScreenshotTest(): image==null")
                    Toaster.show("error: image null 检查日志")
                }
            }
        } catch (e: Exception) {
//            loggerTrace.write("$TAG: takeOneScreenshot: $e \n${exceptionStackTrace(e)}")
            Toaster.show("error: acquireLatestImage() 检查日志")
            return null
        }
        return bitmap
    }

    private fun processBitmap(bitmap: Bitmap) {
        val imageProcessor: ImageProcessor by getKoin().inject()
        val recruitManager: RecruitManager by getKoin().inject()
        val floatingWindow: FloatingWindow by getKoin().inject()

        val result = imageProcessor.getRecruitTags(
            bitmap,
            screenWidth,
            screenHeight
        )
        bitmap.recycle()

        val tags: List<String>
        when (result.status) {
            ImageProcessor.NONE_INFO, ImageProcessor.ERROR_REC, ImageProcessor.UK_RESULT -> {
                Toaster.show(result.msg)
//                if (debug)
//                    loggerRun.write("${result.msg}\n--------end-------")
                return
            }

            else -> {
                tags = result.tags
            }
        }
//        if (debug) {
//            loggerRun.write("tags ${tags[0]} ${tags[1]} ${tags[2]} ${tags[3]} ${tags[4]}")
//            loggerRun.write("--------end----------")
//        }
        Log.e("processBitmap tag size", tags.size.toString())
        val tagsList = getEleCombination(tags)
        val recruitResultList = mutableListOf<RecruitManager.RecruitResult>()
        for (tagsCom in tagsList) {
            val recruitResult = recruitManager.getRecruitResult(tagsCom, true)
            if (recruitResult.operators.isNotEmpty()) {
                recruitResult.sort()
                recruitResultList.add(recruitResult)
            }
        }

        val finalList: List<RecruitManager.RecruitResult> = recruitResultList.toList().sorted()
        Log.e("processBitmap finalList size", finalList.size.toString())
        Log.e("processBitmap finalList size", prefManager.recruitShowMode.get().toString())
        when (prefManager.recruitShowMode.get()) {
            PrefManager.ShowMode.TOAST -> {
                showToastResult(finalList)
            }

            null, PrefManager.ShowMode.FLOAT -> {
                Handler(Looper.getMainLooper()).post {
                    floatingWindow.openAndUpdateWindow(this, tags, finalList)
                }
            }

            PrefManager.ShowMode.AUTO -> {
                if (finalList.isNotEmpty()) {
                    when (finalList[0].rare) {
                        1, 5, 6 -> {
                            Handler(Looper.getMainLooper()).post {
                                floatingWindow.openAndUpdateWindow(this, tags, finalList)
                            }
                        }

                        else -> {
                            showToastResult(finalList)
                        }
                    }
                } else {
                    showToastResult(finalList)
                }
            }
        }
    }

    private fun showToastResult(result: List<RecruitManager.RecruitResult>) {
        Log.e("showToastResult", result.size.toString())
        if (result.isEmpty()) {
            Toaster.show("无四星以上组合")
            return
        } else {
            val item = result[0]
            val str = StringBuilder()

            val rare = item.rare
            val tags = item.tags
            for (tag in tags) {
                str.append(tag).append(" ")
            }
            str.append("-> ").append(rare).append("★").append(" ")
            for (ope in item.operators) {
                if (ope.rare == rare) {
                    str.append(ope.name)
                    break
                }
            }
            Toaster.show(str)
        }

    }

    private fun prepareForScreenshot(): Boolean {
        try {
            imageReader = ImageReader.newInstance(
                screenWidth,
                screenHeight, PixelFormat.RGBA_8888, 2
            )
            surface = imageReader!!.surface
//            if (debug)
//                loggerRun.write("set ImageReader and surface done.")

            mediaProjection = APP.createMediaProjection()
//            if (debug)
//                loggerRun.write("create MediaProjection done")

            startVirtualDisplay()
//            if (debug)
//                loggerRun.write("start VirtualDisplay ")
        } catch (e: Exception) {
            //loggerTrace.write("$TAG: prepareForScreenshot(): $e \n${exceptionStackTrace(e)}")
            Toaster.show("error: prepareForScreenshot 检查日志")
            e.printStackTrace()
            return false
        }
        return true
    }

    private fun startVirtualDisplay() {
        virtualDisplay?.release()
        virtualDisplay = mediaProjection!!.createVirtualDisplay(
            "ScreenShot",
            screenWidth,
            screenHeight,
            screenDensityDpi,
            DisplayManager.VIRTUAL_DISPLAY_FLAG_OWN_CONTENT_ONLY
                    or DisplayManager.VIRTUAL_DISPLAY_FLAG_PUBLIC,
            surface,
            null,
            null
        )
    }

    private fun stop() {
        stopVirtualDisplay()
        stopSurface()
        imageReader = null
        mediaProjection?.stop()
        mediaProjection = null
        APP.stopScreenTaskService()
        finish()
    }

    private fun stopVirtualDisplay() {
        virtualDisplay?.release()
        virtualDisplay = null
    }

    private fun stopSurface() {
        surface?.release()
        surface = null
    }

    private fun releaseMutex() {
        APP.releaseMutexActivity()
    }

    override fun onDestroy() {
        stop()
        releaseMutex()
        super.onDestroy()
    }

}